File: experiments/76-scientific-specimen-atlas/scripts/verify.mjs
import {readFile,writeFile,stat} from 'node:fs/promises';
import {execFileSync} from 'node:child_process';
import {createHash} from 'node:crypto';
import {fileURLToPath} from 'node:url';
import {join,resolve} from 'node:path';
import assert from 'node:assert/strict';
import {readings} from '../src/content.mjs';
const out=fileURLToPath(new URL('../',import.meta.url)),root=resolve(out,'../..');
const sources=JSON.parse(await readFile(join(out,'evidence/sources.json')));
const observations=JSON.parse(await readFile(join(out,'evidence/observations.json')));
const ui=JSON.parse(await readFile(join(out,'evidence/browser-checks.json')));
assert.equal(ui.passed,true);
const clean=[];
for(const [subject,pin] of Object.entries(sources.pins)){
 const wt=join(root,'worktrees',subject);
 assert.equal(execFileSync('git',['-C',wt,'rev-parse','HEAD'],{encoding:'utf8'}).trim(),pin);
 assert.equal(execFileSync('git',['-C',wt,'status','--porcelain','--untracked-files=no'],{encoding:'utf8'}).trim(),'');
 clean.push(subject);
}
for(const d of sources.documents){
 const bytes=execFileSync('git',['-C',join(root,'worktrees',d.subject),'show',`${d.revision}:${d.path}`]);
 assert.equal(bytes.toString(),d.text);assert.equal(createHash('sha256').update(bytes).digest('hex'),d.sha256);
}
for(const receipt of sources.receipts.filter(r=>!['pr127','pr131'].includes(r.id))){
 const doc=sources.documents.find(d=>d.subject===receipt.subject&&d.path===receipt.path);
 assert.equal(doc.text.split(/\r?\n/).slice(receipt.start-1,receipt.end).join('\n'),receipt.text);
}
for(const r of readings)for(const id of r.sources)assert.ok(sources.receipts.some(s=>s.id===id));
for(const name of ['index.html','evidence.html']){
 const html=await readFile(join(out,name),'utf8');
 for(const match of html.matchAll(/(?:href|src)="([^"]+)"/g)){
  const target=match[1];if(/^(https?:|data:)/.test(target))continue;
  const [file,anchor]=target.split('#');
  const destination=join(out,file||name);assert.ok(await stat(destination));
  if(anchor){const text=await readFile(destination,'utf8');assert.ok(text.includes(`id="${anchor}"`),target)}
 }
}
assert.equal(observations.cache.length,10);assert.equal(observations.delivery.length,8);assert.equal(observations.outer.length,4);
assert.deepEqual(observations.factory,{undefinedCalls:1,throwCalls:2,rejectionCalls:1});
assert.deepEqual(sources.inventories['127'],{changedFiles:6,additions:391,deletions:67,originalFiveTestsAndHelpersUnchanged:true,newServiceCases:6,newCoreCases:4});
const tests127=await readFile(join(out,'evidence/tests-127-head.txt'),'utf8');
const tests131=await readFile(join(out,'evidence/tests-131-head.txt'),'utf8');
assert.match(tests127,/Tests\s+15 passed/);assert.match(tests131,/Tests\s+35 passed/);
const readme=await readFile(join(out,'README.md'),'utf8');
for(const heading of ['Entry point','Kind','Subjects','Declared choices','What I tried','What I would drop','What I would do next','Time spent'])assert.ok(readme.includes(`## ${heading}`));
const result={passed:true,cleanTrackedWorktrees:clean,frozenDocuments:sources.documents.length,receipts:sources.receipts.length,findings:readings.length,sourceHashesAndExcerptsMatch:true,localLinksAndAnchorsExist:true,recordedCases:{cache:10,sockets:8,outerLoop:4,factoryOutcomes:3},focusedHeadTests:{pr127:15,pr131:35},readmeContract:true};
await writeFile(join(out,'evidence/static-checks.json'),JSON.stringify(result,null,2)+'\n');console.log(JSON.stringify(result,null,2));

File: experiments/69-command-me-reader/scripts/verify.py
from pathlib import Path
from html.parser import HTMLParser
from urllib.parse import urlsplit, unquote
import json, hashlib, subprocess, re

OUT=Path(__file__).resolve().parents[1]
ROOT=OUT.parents[1]
class Page(HTMLParser):
    def __init__(self,text):
        super().__init__();self.ids=set();self.links=[];self.feed(text)
    def handle_starttag(self,tag,attrs):
        attrs=dict(attrs)
        if 'id' in attrs:
            assert attrs['id'] not in self.ids,attrs['id']
            self.ids.add(attrs['id'])
        for key in ('href','src'):
            if attrs.get(key):self.links.append(attrs[key])

pages={name:Page((OUT/name).read_text()) for name in ['index.html','sources.html']}
checked_links=0
for name,page in pages.items():
    for link in page.links:
        url=urlsplit(link)
        if url.scheme or url.netloc:continue
        path=unquote(url.path) or name
        assert (OUT/path).is_file(),(name,link)
        if url.fragment and path in pages:assert unquote(url.fragment) in pages[path].ids,(name,link)
        checked_links+=1
docs=json.loads((OUT/'evidence/sources.json').read_text())
receipts=json.loads((OUT/'evidence/receipts.json').read_text())
for key,d in docs.items():
    assert hashlib.sha256(d['text'].encode()).hexdigest()==d['sha256']
    tree=ROOT/'worktrees'/('pr-127-'+key.split(':')[0])
    actual=subprocess.check_output(['git','show',d['revision']+':'+d['path']],cwd=tree).decode()
    assert d['text']==actual,key
for key,r in receipts.items():
    if 'document' not in r:continue
    assert r['text']=='\n'.join(docs[r['document']]['text'].splitlines()[r['start']-1:r['end']]),key
assert 'TurnScopedCacheScope' in receipts['public']['text']
assert 'releaseTransientResources(): void' in receipts['base-release']['text']
assert 'does not let an old promise settlement replace a new turn entry' in receipts['core-tests']['text']
inventory=json.loads((OUT/'evidence/inventory.json').read_text())
assert (inventory['changedFiles'],inventory['addedLines'],inventory['removedLines'])==(6,391,67)
assert inventory['originalTestsAndHelpersByteIdentical']
decisions=json.loads((OUT/'src/decisions.json').read_text())
ids={d['id'] for d in decisions}
file_map={
 'typescript-backend.ts':['refresh','release','projections'],
 'typescript-semantic-query-service.ts':['ownership','identity','held','projections','refresh','release','surface'],
 'typescript-semantic-query-service.test.ts':['identity','projections','refresh','release','tests'],
 'turn-scoped-cache-scope.ts':['ownership','identity','held','surface'],
 'turn-scoped-cache-scope.test.ts':['identity','held','tests'],
 'index.ts':['surface'],
}
patch=(OUT/'evidence/diff.patch').read_text()
coverage=[]
for part in patch.split('diff --git ')[1:]:
    path=part.splitlines()[0].split(' b/')[1]
    readings=file_map[Path(path).name];assert all(x in ids for x in readings)
    coverage.append({'path':path,'hunks':re.findall(r'^@@.*@@.*$',part,re.M),'openReadings':readings})
stage_map={'point':['identity'],'drag':['identity','held'],'predict':['held'],'hold':['held'],'holding':['held'],'compare':['held','tests'],'say':['held']}
assert all(x in ids for values in stage_map.values() for x in values)
(OUT/'evidence/coverage.json').write_text(json.dumps({'completeLayer':'opening, instrument legend, ownership strip and eight open readings','changedPaths':coverage,'ritualStages':stage_map,'sourceLinks':{d['id']:d['sources'] for d in decisions}},indent=2)+'\n')
status={}
for build,pin in inventory['pins'].items():
    tree=ROOT/'worktrees'/('pr-127-'+build)
    assert subprocess.check_output(['git','rev-parse','HEAD'],cwd=tree,text=True).strip()==pin
    status[build]=subprocess.check_output(['git','status','--porcelain','--untracked-files=no'],cwd=tree,text=True)
    assert status[build]=='',status
required=['Entry point','Kind','Subjects','Declared choices','What I tried','What I would drop','What I would do next','Time spent']
readme=(OUT/'README.md').read_text()
for h in required:assert '## '+h in readme,h
assert (OUT/'brief.md').read_text()==(ROOT/'queue/running/69-command-me-reader.md').read_text() if (ROOT/'queue/running/69-command-me-reader.md').exists() else (OUT/'brief.md').is_file()
assert '15 passed' in (OUT/'evidence/head-tests.txt').read_text()
browser=json.loads((OUT/'validation-browser.json').read_text());assert browser['passed']
result={'passed':True,'localLinksChecked':checked_links,'sourceDocumentsVerified':len(docs),'receiptsVerified':len(receipts),'completeReadings':len(decisions),'changedPathsCovered':len(coverage),'hunksCovered':sum(len(x['hunks']) for x in coverage),'stagesCovered':len(stage_map),'trackedWorktreeStatus':status,'builtAssets':{p:hashlib.sha256((OUT/p).read_bytes()).hexdigest() for p in ['index.html','sources.html','app.js','style.css']}}
(OUT/'validation-static.json').write_text(json.dumps(result,indent=2)+'\n')
print(json.dumps(result,indent=2))
